libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/Plugins/SFNetworkPlugin/SFNetworkPlugin.m

```diff
@@ -0,0 +1,120 @@
+NSString * const kFileUrl         = @"fileUrl";
+NSString * const kFileName        = @"fileName";
+NSString * const kPatchMethod     = @"PATCH";
+
+@implementation SFNetworkPlugin
+
+- (void) pgSendRequest:(CDVInvokedUrlCommand *) command
+{
+    NSDictionary *argsDict = [self getArgument:command.arguments atIndex:0];
+    SFRestMethod method = [SFRestRequest sfRestMethodFromHTTPMethod:[argsDict nonNullObjectForKey:kMethodArg]];
+    NSString* endPoint = [argsDict nonNullObjectForKey:kEndPointArg];
+    NSString* path = [argsDict nonNullObjectForKey:kPathArg];
+
+    /*
+     * Android's network library has a limitation that it does not support
+     * PATCH requests. However, Salesforce REST API does not allow POST for
+     * updates, it only allows PATCH. Hence, we work around the issue on
+     * Android by passing in method name PATCH as a URL parameter. However,
+     * we don't need this on iOS, since we can directly make a PATCH request.
+     */
+    if ([path containsString:kPatchMethod]) {
+        method = SFRestMethodPATCH;
+    }
+    NSDictionary* queryParams = [[NSDictionary alloc] init];
+    id queryParamsObj = [argsDict nonNullObjectForKey:kQueryParams];
+
+    /*
+     * Query params are NSDictionary for GET and encoded JSON in NSString
+     * for POST. These checks ensure that both cases are handled properly.
+     */
+    if ([queryParamsObj isKindOfClass:[NSString class]]) {
+        NSData* queryParamsData = [queryParamsObj dataUsingEncoding:NSUTF8StringEncoding];
+        queryParams = [NSJSONSerialization JSONObjectWithData:queryParamsData options:0 error:nil];
+    } else {
+        queryParams = queryParamsObj;
+    }
+    NSDictionary* headerParams = [argsDict nonNullObjectForKey:kHeaderParams];
+    NSDictionary* fileParams = [argsDict nonNullObjectForKey:kfileParams];
+    SFRestRequest* request = [SFRestRequest requestWithMethod:method path:path queryParams:queryParams];
+
+    // Adds custom headers, if any.
+    [request setCustomHeaders:headerParams];
+    if (endPoint) {
+        [request setEndpoint:endPoint];
+    }
+
+    // Sets body for a file POST request.
+    if (fileParams) {
+
+        /*
+         * File params expected to be of the form:
+         * {<fileParamNameInPost>: {fileMimeType:<someMimeType>, fileUrl:<fileUrl>, fileName:<fileNameForPost>}}.
+         */
+        for (NSString* fileParamName in fileParams) {
+            NSDictionary* fileParam = fileParams[fileParamName];
+            NSString* fileMimeType = [fileParam nonNullObjectForKey:kFileMimeType];
+            NSString* fileUrl = [fileParam nonNullObjectForKey:kFileUrl];
+            NSString* fileName = [fileParam nonNullObjectForKey:kFileName];
+            NSData* fileData = [NSData dataWithContentsOfURL:[NSURL URLWithString:fileUrl]];
+            [request addPostFileData:fileData paramName:fileParamName fileName:fileName mimeType:fileMimeType];
+        }
+    }
+    [[SFRestAPI sharedInstance] sendRESTRequest:request
```

Comment:
This needs a weakify/strongify pattern, to avoid capturing self:

```obj-c
__weak typeof(self) weakSelf = self;
[[SFRestAPI sharedInstance] sendRESTRequest:request
                                  failBlock:^(NSError *e) {
                                      __strong typeof(self) strongSelf = weakSelf;
                                      CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:e.localizedDescription];
                                      [strongSelf.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
                                  }
                              completeBlock:^(id response) {
                                  __strong typeof(self) strongSelf = weakSelf;
                                  CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:response];
                                  [strongSelf.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
                              }
 ];
```